options/pin.go

```diff
@@ -86,36 +117,131 @@ func PinUpdateOptions(opts ...PinUpdateOption) (*PinUpdateSettings, error) {
+// type.
+//
+// Supported values:
+// * "direct" - directly pinned objects
+// * "recursive" - roots of recursive pins
+// * "indirect" - indirectly pinned objects (referenced by recursively pinned
+//    objects)
+// * "all" - all pinned objects (default)
+func (pinLsOpts) Type(typeStr string) (PinLsOption, error) {
+	switch typeStr {
+	case "all", "direct", "indirect", "recursive":
+		return Pin.Ls.pinType(typeStr), nil
+	default:
+		return nil, fmt.Errorf("invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
+	}
+}
+
+// pinType is an option for Pin.Ls which allows to specify which pin types should
+// be returned
+//
+// Supported values:
+// * "direct" - directly pinned objects
+// * "recursive" - roots of recursive pins
+// * "indirect" - indirectly pinned objects (referenced by recursively pinned
+//    objects)
+// * "all" - all pinned objects (default)
+func (pinLsOpts) pinType(t string) PinLsOption {
+	return func(settings *PinLsSettings) error {
+		settings.Type = t
+		return nil
+	}
+}
+
+type pinIsPinnedOpts struct{}
+
+// All is an option for Pin.IsPinned which will make it search in all type of pins.
+// It is the default
+func (pinIsPinnedOpts) All() PinIsPinnedOption {
+	return Pin.IsPinned.pinType("all")
+}
+
+// Recursive is an option for Pin.IsPinned which will make it only search in
+// recursive pins
+func (pinIsPinnedOpts) Recursive() PinIsPinnedOption {
+	return Pin.IsPinned.pinType("recursive")
+}
+
+// Direct is an option for Pin.IsPinned which will make it only search in direct
+// (non recursive) pins
+func (pinIsPinnedOpts) Direct() PinIsPinnedOption {
+	return Pin.IsPinned.pinType("direct")
+}
+
+// Indirect is an option for Pin.IsPinned which will make it only search indirect
+// pins (objects referenced by other recursively pinned objects)
+func (pinIsPinnedOpts) Indirect() PinIsPinnedOption {
+	return Pin.IsPinned.pinType("indirect")
+}
+
+// Type is an option for Pin.IsPinned which will make it only search pins of the given
+// type.
+//
+// Supported values:
+// * "direct" - directly pinned objects
+// * "recursive" - roots of recursive pins
+// * "indirect" - indirectly pinned objects (referenced by recursively pinned
+//    objects)
+// * "all" - all pinned objects (default)
+func (pinIsPinnedOpts) Type(typeStr string) (PinIsPinnedOption, error) {
+	switch typeStr {
+	case "all", "direct", "indirect", "recursive":
+		return Pin.IsPinned.pinType(typeStr), nil
+	default:
+		return nil, fmt.Errorf("invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
+	}
+}
+
+// pinType is an option for Pin.IsPinned which allows to specify which pin type the given
+// pin is expected to be, speeding up the research.
+//
+// Supported values:
+// * "direct" - directly pinned objects
+// * "recursive" - roots of recursive pins
+// * "indirect" - indirectly pinned objects (referenced by recursively pinned
+//    objects)
+// * "all" - all pinned objects (default)
```

Comment:
This does not really apply, since this one does not verify anything. Having a copy-paste from the `Type` docstring is confusing.

It seems this is not part of `Type` just to avoid having an extra line in `All()`, `Recursive()` etc. It seems clearer to me to simply handle everything inside `Type`.

Reply:
Not sure if you want more change about this. Maybe something like this ?

```go
type pinLsOpts struct{}

// All is an option for Pin.Ls which will make it return all pins. It is
// the default
func (pinLsOpts) All() PinLsOption {
	opt, _ := Pin.Ls.Type("all")
	return opt
}

// Recursive is an option for Pin.Ls which will make it only return recursive
// pins
func (pinLsOpts) Recursive() PinLsOption {
	opt, _ := Pin.Ls.Type("recursive")
	return opt
}

// Direct is an option for Pin.Ls which will make it only return direct (non
// recursive) pins
func (pinLsOpts) Direct() PinLsOption {
	opt, _ := Pin.Ls.Type("direct")
	return opt
}

// Indirect is an option for Pin.Ls which will make it only return indirect pins
// (objects referenced by other recursively pinned objects)
func (pinLsOpts) Indirect() PinLsOption {
	opt, _ := Pin.Ls.Type("indirect")
	return opt
}

// Type is an option for Pin.Ls which will make it only return pins of the given
// type.
//
// Supported values:
// * "direct" - directly pinned objects
// * "recursive" - roots of recursive pins
// * "indirect" - indirectly pinned objects (referenced by recursively pinned
//    objects)
// * "all" - all pinned objects (default)
func (pinLsOpts) Type(typeStr string) (PinLsOption, error) {
	switch typeStr {
	case "all", "direct", "indirect", "recursive":
		return func(settings *PinLsSettings) error {
			settings.Type = typeStr
			return nil
		}, nil
	default:
		return nil, fmt.Errorf("invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
	}
}
```

Reply:
... and remove the "internal" `pinType`